applications/OptimizationApplication/python_scripts/controls/material/material_properties_control.py

```diff
@@ -0,0 +1,89 @@
+import KratosMultiphysics as Kratos
+import KratosMultiphysics.OptimizationApplication as KratosOA
+from KratosMultiphysics.OptimizationApplication.controls.control import Control
+from KratosMultiphysics.OptimizationApplication.utilities.optimization_info import OptimizationInfo
+from KratosMultiphysics.OptimizationApplication.utilities.union_utilities import SupportedSensitivityFieldVariableTypes
+
+class MaterialPropertiesControl(Control):
+    """Material properties control
+
+    This is a generic material properties control which creates a control
+    for the specified control variable. This does not do any filtering.
+
+    TODO: Extend with filtering techniques when they are implemented.
+
+    """
+    def __init__(self, model: Kratos.Model, parameters: Kratos.Parameters, optimization_info: OptimizationInfo):
+        super().__init__()
+
+        default_settings = Kratos.Parameters("""{
+            "model_part_names"            : [""],
+            "control_variable_name"       : ""
+        }""")
+        parameters.ValidateAndAssignDefaults(default_settings)
+
+        self.optimization_info = optimization_info
+        self.model_parts = [model[model_part_name] for model_part_name in parameters["model_part_names"].GetStringArray()]
+
+        control_variable_name = parameters["control_variable_name"].GetString()
+        control_variable_type = Kratos.KratosGlobals.GetVariableType(control_variable_name)
+        if control_variable_type != "Double":
+            raise RuntimeError(f"{control_variable_name} with {control_variable_type} type is not supported. Only supports double variables")
+
+        sensitivity_variable_name = control_variable_name + "_SENSITIVITY"
```

Comment:
I understand this but I don't have a solution for the following question.... How can the algorithm know which control requires which sensitivities to be taken from response function?

Current situation is:
We have `Response`, `Control` and `Algorithm`. Algorithm needs to know which sensitivities to be requested from response which is need to be passed to control. How can it know? Response only knows all the types of variables which it has dependence on. Control knows about the updating and its mechanisms, and what it requires to update. Thats why in current design I put required sensitivities in `Control`.

Following are the alternatives which came to my mind, (none of them are perfect).

My solutions:
1. Having a map of control variables and sensitivities (a global map) -> I don't like maintaining maps (error prone, hard to debug)
2. Requesting from control what sensitivities are required from their side. -> Current design
3. Passing the response as an input to the control so control can calculate what ever is required from the sensitivities. -> What I have proposed previously. -> I don't like the idea of passing resposne to the control. (Interface becomes entangled).

I am really open for suggestions since none of my solutions are the ideal.

Reply:
We need to discuss this as explained earlier. I don't like this either, but this was the least bad solution which I thought.